sound/soc/sof/sof-audio.c

```diff
@@ -27,31 +27,6 @@ static int sof_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_sof_control *
 	return ret;
 }
 
-static int sof_dai_config_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *dai)
-{
-	struct sof_dai_private_data *private = dai->private;
-	struct sof_ipc_dai_config *config;
-	struct sof_ipc_reply reply;
-	int ret;
-
-	config = &private->dai_config[dai->current_config];
-	if (!config) {
-		dev_err(sdev->dev, "error: no config for DAI %s\n", dai->name);
-		return -EINVAL;
-	}
-
-	/* set NONE flag to clear all previous settings */
-	config->flags = SOF_DAI_CONFIG_FLAGS_NONE;
-
-	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
-				 &reply, sizeof(reply));
-
-	if (ret < 0)
-		dev_err(sdev->dev, "error: failed to set dai config for %s\n", dai->name);
-
-	return ret;
-}
-
```

Comment:
Do you have idea if we could move out the content of `sof_widget_kcontrol_setup()` (along with the `sof_kcontrol_setup()`) to somewhere (ipc3-control.c?) to remove the snd_sof_ipc_set_get_comp_data()?
As an ops with scontrol parameter perhaps?

Reply:
Can it be blended into tplg_ops->widget_setup?

Reply:
we need sof_widget_kcontrol_setup() here isnt? because even for IPC4, we need to restore the kcontrol data when we set up the widget, right?

Reply:
once you have your set_get_data op, we can simply use that op in sof_kcontrol_set() maybe

Reply:
@ranj063, the set_get_data is not that simple and we will need two level here:
set_get_control_data() needs to be different for IPC3 and IPC4 to construct the message to be sent from the control.
then
set_get_data(), again IPC3 and IPC4 version. This knows how to send a message in chunks.

The reason is that in IPC3 _only_ `SOF_IPC_GLB_COMP_MSG` can be split up, no other message is supported, while in IPC4 other types of messages can be split up as well.

In IPC4 we would have ipc4_set_get_control_data() to call the ipc4_set_get_data() and possible other paths which would also call the same set_get_data(), but for different message types.

Reply:
Looking at this again, imho the feature you re referring is broken.
`sof_widget_kcontrol_setup()` will first **set** the data to the firmware then, if the swidget is not part of a dynamic pipeline it will *get** the data back from the firmware.
So we read back what we just wrote?

Reply:
I'll take care of this one in a separate PR where it can be done a bit cleaner way.